kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/PortConflictDetector.java

```diff
@@ -8,242 +8,112 @@
-        return ports.stream().filter(candidates::contains).sorted().toList();
+    /**
+     * Validates the configuration throwing an exception if a conflict is detected.
+     *
+     * @param virtualClusterModelMap map of virtual clusters.
+     * @param otherExclusivePort an optional exclusive port that should conflict with virtual cluster ports
+     */
+    public void validate(Collection<VirtualClusterModel> virtualClusterModelMap, Optional<HostPort> otherExclusivePort) {
+        List<CandidateBinding> candidateBindings = candidateBindings(virtualClusterModelMap);
+        validateCandidatesDoNotConflictWithOtherExclusivePort(candidateBindings, otherExclusivePort);
+        validateCandidatesDoNotConflictWithEachOther(candidateBindings);
     }
 
-    private IllegalStateException buildPortConflictException(EndpointListener listener,
-                                                             Set<String> seenVirtualClusters,
-                                                             List<Integer> conflicts,
-                                                             Optional<String> proposedBindingInterface,
-                                                             BindingScope proposedScope,
-                                                             Optional<String> existingBindingInterface,
-                                                             BindingScope existingBindingScope) {
-        var portConflicts = conflicts.stream().map(String::valueOf).collect(Collectors.joining(","));
-
-        var underlying = new IllegalStateException("The %s bind of port(s) %s to %s would conflict with existing %s port bindings on %s.".formatted(
-                proposedScope.name().toLowerCase(Locale.ROOT),
-                portConflicts,
-                proposedBindingInterface.orElse(ANY_STRING),
-                existingBindingScope.name().toLowerCase(Locale.ROOT),
-                existingBindingInterface.orElse(ANY_STRING)));
-        return buildOverviewException(listener, seenVirtualClusters, underlying);
+    private static void validateCandidatesDoNotConflictWithOtherExclusivePort(List<CandidateBinding> candidateBindings, Optional<HostPort> otherExclusivePort) {
+        otherExclusivePort.ifPresent((hostPort) -> {
+            candidateBindings.forEach(c -> c.validateNonConflicting(hostPort));
+        });
     }
 
-    private IllegalStateException buildTlsConflictException(Integer port, Optional<String> bindingAddress) {
-        return new IllegalStateException(
-                "The shared bind of port %d to %s has conflicting TLS settings with existing port on the same interface.".formatted(port,
-                        bindingAddress.orElse(ANY_STRING)));
+    private static void validateCandidatesDoNotConflictWithEachOther(List<CandidateBinding> candidateBindings) {
+        // compare all unique combinations
+        for (int i = 0; i < candidateBindings.size(); i++) {
```

Comment:
could be changed to `i < candidateBindings.size() - 1` since we don't need to compare the final item to anything, but I think it makes it harder to read.